File: cli/llm_sanitize.py
"""Санитизация сырого текста от локальных моделей (Gemma thinking, псевдо-tool_call).

Некоторые OpenAI-совместимые прокси отдают в ``message.content`` внутренние каналы
(``<|channel>thought``) и вызовы инструментов текстом (``<|tool_call>call:…``),
без поля ``tool_calls``. CLI приводит это к ожидаемому формату chat API.

Отключить: ``EIDOS_STRIP_MODEL_THOUGHT=0`` (по умолчанию включено).
"""

from __future__ import annotations

import json
import os
import re
import uuid
from typing import Any

_THOUGHT_OPEN = re.compile(r"<\|channel>thought\b", re.IGNORECASE)
_CHANNEL_OR_TOOL = re.compile(
    r"<\|(?:channel>(?:final|comment|response)|tool_call>|think\|>)",
    re.IGNORECASE,
)
_CHANNEL_FINAL_PREFIX = re.compile(
    r"<\|channel>(?:final|comment|response)\s*",
    re.IGNORECASE,
)
_JUNK_TOKENS = (
    re.compile(r"<\|\"\|>", re.IGNORECASE),
    re.compile(r"<\|think\|>", re.IGNORECASE),
    re.compile(r"<\|end\|>", re.IGNORECASE),
)
_TOOL_CALL_MARKER = "<|tool_call>call:"


def strip_model_thought_enabled() -> bool:
    v = os.environ.get("EIDOS_STRIP_MODEL_THOUGHT", "1").strip().lower()
    return v not in ("0", "false", "no", "off")


def strip_model_channels(text: str) -> str:
    """Убрать блоки ``<|channel>thought`` и служебные токены; оставить видимый ответ."""
    out = text
    while True:
        m = _THOUGHT_OPEN.search(out)
        if not m:
            break
        start = m.start()
        tail = out[m.end() :]
        end_m = _CHANNEL_OR_TOOL.search(tail)
        if end_m:
            out = out[:start] + tail[end_m.start() :]
        else:
            out = out[:start]
            break
    out = _CHANNEL_FINAL_PREFIX.sub("", out)
    inline = _iter_gemma_inline_tool_calls(out)
    out = _remove_inline_tool_call_spans(out, inline)
    for pat in _JUNK_TOKENS:
        out = pat.sub("", out)
    return out.strip()


def _parse_pseudo_tool_args(raw: str) -> dict[str, Any]:
    """Разбор ``max_chars:1,selector:…`` из Gemma-текста в dict для JSON arguments."""
    s = raw.replace('<|"|>', '"').strip()
    out: dict[str, Any] = {}
    m_chars = re.match(r"max_chars\s*:\s*(\d+)", s)
    if m_chars:
        out["max_chars"] = int(m_chars.group(1))
        s = s[m_chars.end() :].lstrip(", ")
    m_sel = re.search(r"selector\s*:\s*(.+)\s*$", s, re.DOTALL)
    if m_sel:
        val = m_sel.group(1).strip()
        if len(val) >= 2 and val[0] == val[-1] == '"':
            val = val[1:-1]
        out["selector"] = val
        return out
    m_url = re.search(r"url\s*:\s*(.+)\s*$", s, re.DOTALL)
    if m_url:
        val = m_url.group(1).strip().strip('"')
        out["url"] = val
        return out
    if s.startswith("{"):
        try:
            parsed = json.loads(s)
            return parsed if isinstance(parsed, dict) else {}
        except json.JSONDecodeError:
            pass
    for part in re.split(r",(?=\w+\s*:)", s):
        if ":" not in part:
            continue
        key, val = part.split(":", 1)
        key = key.strip()
        val = val.strip().strip('"').strip("'")
        if key == "max_chars":
            try:
                out[key] = int(val)
            except ValueError:
                out[key] = val
        elif val:
            out[key] = val
    return out


def _iter_gemma_inline_tool_calls(text: str) -> list[tuple[str, str, int, int]]:
    """Найти ``<|tool_call>call:name{args}<tool_call|>`` с балансом фигурных скобок."""
    found: list[tuple[str, str, int, int]] = []
    pos = 0
    while True:
        idx = text.find(_TOOL_CALL_MARKER, pos)
        if idx < 0:
            break
        name_start = idx + len(_TOOL_CALL_MARKER)
        brace = text.find("{", name_start)
        if brace < 0:
            break
        name = text[name_start:brace].strip()
        depth = 0
        j = brace
        end_brace = -1
        while j < len(text):
            ch = text[j]
            if ch == "{":
                depth += 1
            elif ch == "}":
                depth -= 1
                if depth == 0:
                    end_brace = j
                    break
            j += 1
        if end_brace < 0:
            break
        args_raw = text[brace + 1 : end_brace]
        close_m = re.match(
            r"\s*(?:<tool_call\|>|<\|tool_call\|>)",
            text[end_brace + 1 :],
            re.IGNORECASE,
        )
        if not close_m:
            pos = end_brace + 1
            continue
        span_end = end_brace + 1 + close_m.end()
        found.append((name, args_raw, idx, span_end))
        pos = span_end
    return found


def _to_openai_tool_calls(
    inline: list[tuple[str, str, int, int]],
) -> list[dict[str, Any]]:
    out: list[dict[str, Any]] = []
    for name, args_raw, _start, _end in inline:
        args = _parse_pseudo_tool_args(args_raw)
        out.append(
            {
                "id": f"call_{uuid.uuid4().hex[:12]}",
                "type": "function",
                "function": {
                    "name": name,
                    "arguments": json.dumps(args, ensure_ascii=False),
                },
            }
        )
    return out


def _remove_inline_tool_call_spans(text: str, spans: list[tuple[str, str, int, int]]) -> str:
    if not spans:
        return text
    parts: list[str] = []
    last = 0
    for _n, _a, start, end in spans:
        parts.append(text[last:start])
        last = end
    parts.append(text[last:])
    return "".join(parts)


def sanitize_assistant_text(text: str) -> tuple[str, list[dict[str, Any]]]:
    """Очистить content и извлечь inline tool_calls, если есть."""
    inline = _iter_gemma_inline_tool_calls(text)
    body = _remove_inline_tool_call_spans(text, inline)
    body = strip_model_channels(body)
    tool_calls = _to_openai_tool_calls(inline)
    return body, tool_calls


def normalize_assistant_message(msg: dict[str, Any]) -> dict[str, Any]:
    """Привести assistant ``message`` к формату OpenAI chat (content / tool_calls)."""
    if not strip_model_thought_enabled():
        return msg
    if msg.get("tool_calls"):
        content = msg.get("content")
        if isinstance(content, str) and (
            "<|channel" in content or "<|tool_call>" in content
        ):
            cleaned, _extra = sanitize_assistant_text(content)
            out = dict(msg)
            out["content"] = cleaned if cleaned else None
            return out
        return msg

    content = msg.get("content")
    if not isinstance(content, str) or not content.strip():
        return msg
    if "<|channel" not in content and "<|tool_call>" not in content and "<|think" not in content:
        return msg

    cleaned, parsed = sanitize_assistant_text(content)
    out = dict(msg)
    if parsed:
        out["tool_calls"] = parsed
        out["content"] = cleaned if cleaned else None
    else:
        out["content"] = cleaned
    return out


def sanitize_assistant_content_for_history(text: str) -> str:
    """Только текст для истории WM (без подъёма tool_calls из старых событий)."""
    if not strip_model_thought_enabled():
        return text
    if "<|channel" not in text and "<|tool_call>" not in text and "<|think" not in text:
        return text
    cleaned, _ = sanitize_assistant_text(text)
    return cleaned

File: logos-rs/crates/eidos-core/src/llm_sanitize.rs
//! Санитизация ответов локальных моделей (Gemma thinking, inline tool_call).

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn strips_thought_block() {
        let raw = "<|channel>thought\nsecret\n<|channel>final\nvisible";
        assert_eq!(strip_model_channels(raw), "visible");
    }
}

use std::env;

use eidos_protocol::working::{FunctionCall, ToolCall};
use regex::Regex;
use serde_json::{json, Value};

const TOOL_CALL_MARKER: &str = "<|tool_call>call:";

fn strip_enabled() -> bool {
    let v = env::var("EIDOS_STRIP_MODEL_THOUGHT")
        .unwrap_or_else(|_| "1".into())
        .to_ascii_lowercase();
    !matches!(v.as_str(), "0" | "false" | "no" | "off")
}

fn thought_open_re() -> Regex {
    Regex::new(r"(?i)<\|channel>thought\b").expect("regex")
}

fn channel_or_tool_re() -> Regex {
    Regex::new(r"(?i)<\|(?:channel>(?:final|comment|response)|tool_call>|think\|>)").expect("regex")
}

fn channel_final_re() -> Regex {
    Regex::new(r"(?i)<\|channel>(?:final|comment|response)\s*").expect("regex")
}

fn tool_close_re() -> Regex {
    Regex::new(r"(?i)\s*(?:<tool_call\|>|<\|tool_call\|>)").expect("regex")
}

/// Убрать блоки thought и служебные токены.
pub fn strip_model_channels(text: &str) -> String {
    let thought_open = thought_open_re();
    let channel_or_tool = channel_or_tool_re();
    let channel_final = channel_final_re();

    let mut out = text.to_string();
    loop {
        let Some(m) = thought_open.find(&out) else {
            break;
        };
        let start = m.start();
        let tail = &out[m.end()..];
        if let Some(end_m) = channel_or_tool.find(tail) {
            out = format!("{}{}", &out[..start], &tail[end_m.start()..]);
        } else {
            out.truncate(start);
            break;
        }
    }
    out = channel_final.replace_all(&out, "").into_owned();
    let inline = iter_gemma_inline_tool_calls(&out);
    out = remove_inline_spans(&out, &inline);
    for junk in ["<|\"|>", "<|think|>", "<|end|>"] {
        out = out.replace(junk, "");
    }
    out.trim().to_string()
}

fn iter_gemma_inline_tool_calls(text: &str) -> Vec<(String, String, usize, usize)> {
    let mut found = Vec::new();
    let mut pos = 0;
    while let Some(idx) = text[pos..].find(TOOL_CALL_MARKER) {
        let idx = pos + idx;
        let name_start = idx + TOOL_CALL_MARKER.len();
        let Some(rel_brace) = text[name_start..].find('{') else {
            break;
        };
        let brace = name_start + rel_brace;
        let name = text[name_start..brace].trim().to_string();
        let mut depth = 0i32;
        let mut end_brace = None;
        for (j, ch) in text[brace..].char_indices() {
            let j = brace + j;
            match ch {
                '{' => depth += 1,
                '}' => {
                    depth -= 1;
                    if depth == 0 {
                        end_brace = Some(j);
                        break;
                    }
                }
                _ => {}
            }
        }
        let Some(end_brace) = end_brace else { break };
        let args_raw = text[brace + 1..end_brace].to_string();
        let close_re = tool_close_re();
        let after = &text[end_brace + 1..];
        let Some(close_m) = close_re.find(after) else {
            pos = end_brace + 1;
            continue;
        };
        let span_end = end_brace + 1 + close_m.end();
        found.push((name, args_raw, idx, span_end));
        pos = span_end;
    }
    found
}

fn remove_inline_spans(text: &str, spans: &[(String, String, usize, usize)]) -> String {
    if spans.is_empty() {
        return text.to_string();
    }
    let mut parts = Vec::new();
    let mut last = 0;
    for (_, _, start, end) in spans {
        parts.push(&text[last..*start]);
        last = *end;
    }
    parts.push(&text[last..]);
    parts.concat()
}

fn parse_pseudo_tool_args(raw: &str) -> Value {
    let s = raw.replace("<|\"|>", "\"");
    let mut out = serde_json::Map::new();
    if let Some(caps) = regex::Regex::new(r"max_chars\s*:\s*(\d+)")
        .unwrap()
        .captures(&s)
    {
        if let Some(m) = caps.get(1) {
            if let Ok(n) = m.as_str().parse::<i64>() {
                out.insert("max_chars".into(), json!(n));
            }
        }
    }
    if let Some(caps) = regex::Regex::new(r"(?is)selector\s*:\s*(.+)\s*$")
        .unwrap()
        .captures(&s)
    {
        let mut val = caps.get(1).map(|m| m.as_str().trim()).unwrap_or("").to_string();
        if val.len() >= 2 && val.starts_with('"') && val.ends_with('"') {
            val = val[1..val.len() - 1].to_string();
        }
        out.insert("selector".into(), json!(val));
        return Value::Object(out);
    }
    if s.trim_start().starts_with('{') {
        if let Ok(v) = serde_json::from_str::<Value>(&s) {
            if v.is_object() {
                return v;
            }
        }
    }
    Value::Object(out)
}

fn to_openai_tool_calls(inline: &[(String, String, usize, usize)]) -> Vec<ToolCall> {
    inline
        .iter()
        .map(|(name, args_raw, _, _)| {
            let args = parse_pseudo_tool_args(args_raw);
            let id = format!(
                "call_{:012x}",
                std::time::SystemTime::now()
                    .duration_since(std::time::UNIX_EPOCH)
                    .map(|d| d.as_nanos())
                    .unwrap_or(0)
            );
            let arguments = serde_json::to_string(&args).unwrap_or_else(|_| "{}".into());
            ToolCall {
                id,
                call_type: "function".to_string(),
                function: FunctionCall {
                    name: name.clone(),
                    arguments,
                },
            }
        })
        .collect()
}

/// Очистить content и извлечь inline tool_calls.
pub fn sanitize_assistant_text(text: &str) -> (String, Vec<ToolCall>) {
    let inline = iter_gemma_inline_tool_calls(text);
    let body = remove_inline_spans(text, &inline);
    let body = strip_model_channels(&body);
    let tool_calls = to_openai_tool_calls(&inline);
    (body, tool_calls)
}

/// Нормализовать объект message из API.
pub fn normalize_assistant_message(mut msg: Value) -> Value {
    if !strip_enabled() {
        return msg;
    }
    if msg.get("tool_calls").is_some() {
        if let Some(content) = msg.get("content").and_then(|c| c.as_str()) {
            if content.contains("<|channel") || content.contains("<|tool_call>") {
                let (cleaned, _) = sanitize_assistant_text(content);
                if cleaned.is_empty() {
                    msg["content"] = Value::Null;
                } else {
                    msg["content"] = Value::String(cleaned);
                }
            }
        }
        return msg;
    }
    let Some(content) = msg.get("content").and_then(|c| c.as_str()) else {
        return msg;
    };
    if !content.contains("<|channel") && !content.contains("<|tool_call>") && !content.contains("<|think")
    {
        return msg;
    }
    let (cleaned, parsed) = sanitize_assistant_text(content);
    if !parsed.is_empty() {
        msg["tool_calls"] = serde_json::to_value(&parsed).unwrap_or(Value::Null);
        msg["content"] = if cleaned.is_empty() {
            Value::Null
        } else {
            Value::String(cleaned)
        };
    } else {
        msg["content"] = Value::String(cleaned);
    }
    msg
}

/// Только текст для истории WM.
pub fn sanitize_assistant_content_for_history(text: &str) -> String {
    if !strip_enabled() {
        return text.to_string();
    }
    if !text.contains("<|channel") && !text.contains("<|tool_call>") && !text.contains("<|think") {
        return text.to_string();
    }
    let (cleaned, _) = sanitize_assistant_text(text);
    cleaned
}
